File: skills/test-graph/test_graph/sources/reset_node.py
# /// script
# requires-python = ">=3.10"
# dependencies = ["testgraphsdk"]
#
# [tool.uv.sources]
# testgraphsdk = { path = "../sdk/python", editable = true }
# ///
from __future__ import annotations

from testgraphsdk import NodeResult, NodeSpec, node, reset_node


TARGETS = (
    ("local-preview", "local"),
    ("local-github-action", "github-action"),
    ("aws-preview", "aws"),
)

SPEC = (
    NodeSpec("tg6.lifecycle.python.reset-node")
    .kind("assertion")
    .tags("tg6", "environment", "lifecycle-template", "uv")
    .output("Runtime")
    .output("ResetCases")
)


@node(SPEC)
def main(ctx):
    run_plans = [reset_node(target, backend) for target, backend in TARGETS]
    just_provisioned = [reset_node(target, backend, just_provisioned=True) for target, backend in TARGETS]
    already_reset = [reset_node(target, backend, already_reset=True) for target, backend in TARGETS]

    result = (
        NodeResult.pass_(ctx.node_id)
        .assertion("all_targets_covered", len(run_plans) == len(TARGETS))
        .assertion("normal_reset_runs", all(plan.should_run for plan in run_plans))
        .assertion("normal_reset_uses_reset_action", all(plan.environment_action == "reset" for plan in run_plans))
        .assertion("normal_reset_uses_apply_command", all(plan.tofu_command == ("tofu", "apply", "-auto-approve") for plan in run_plans))
        .assertion("just_provisioned_reset_skips", all(not plan.should_run for plan in just_provisioned))
        .assertion("just_provisioned_reason_recorded", all(plan.skip_reason == "just-provisioned" for plan in just_provisioned))
        .assertion("already_reset_skips", all(not plan.should_run for plan in already_reset))
        .assertion("already_reset_reason_recorded", all(plan.skip_reason == "already-reset" for plan in already_reset))
        .assertion("skip_keeps_environment_active", all(plan.expected_state == "kept-active" for plan in just_provisioned + already_reset))
        .publish("Runtime", "uv")
        .publish("ResetCases", str(len(run_plans) + len(just_provisioned) + len(already_reset)))
    )
    for key, value in reset_node("local-preview", "local", just_provisioned=True).published().items():
        result.publish(f"Reset{key}", value)
    return result


if __name__ == "__main__":
    main()

File: skills/test-graph/test_graph/sources/ResetNode.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//SOURCES ../sdk/java/src/main/java/com/hayden/testgraphsdk/sdk/*.java

import java.util.List;
import java.util.Map;

import com.hayden.testgraphsdk.sdk.ClusterLifecycle;
import com.hayden.testgraphsdk.sdk.ClusterLifecyclePlan;
import com.hayden.testgraphsdk.sdk.Node;
import com.hayden.testgraphsdk.sdk.NodeResult;
import com.hayden.testgraphsdk.sdk.NodeSpec;

public class ResetNode {
    private record Target(String target, String backend) {}

    private static final List<Target> TARGETS = List.of(
            new Target("local-preview", "local"),
            new Target("local-github-action", "github-action"),
            new Target("aws-preview", "aws")
    );

    private static final NodeSpec SPEC = NodeSpec.of("tg6.lifecycle.java.reset-node")
            .kind(NodeSpec.Kind.ASSERTION)
            .tags("tg6", "environment", "lifecycle-template", "jbang")
            .output("Runtime", "string")
            .output("ResetCases", "string");

    public static void main(String[] args) {
        Node.run(args, SPEC, ctx -> {
            boolean normalRuns = true;
            boolean justProvisionedSkips = true;
            boolean alreadyResetSkips = true;
            int cases = 0;
            for (Target target : TARGETS) {
                ClusterLifecyclePlan run = ClusterLifecycle.resetNode(target.target(), target.backend(), false, false);
                ClusterLifecyclePlan justProvisioned = ClusterLifecycle.resetNode(target.target(), target.backend(), true, false);
                ClusterLifecyclePlan alreadyReset = ClusterLifecycle.resetNode(target.target(), target.backend(), false, true);
                cases += 3;
                normalRuns &= run.shouldRun()
                        && "reset".equals(run.environmentAction())
                        && run.tofuCommand().equals(List.of("tofu", "apply", "-auto-approve"));
                justProvisionedSkips &= !justProvisioned.shouldRun()
                        && "just-provisioned".equals(justProvisioned.skipReason())
                        && "kept-active".equals(justProvisioned.expectedState());
                alreadyResetSkips &= !alreadyReset.shouldRun()
                        && "already-reset".equals(alreadyReset.skipReason())
                        && "kept-active".equals(alreadyReset.expectedState());
            }

            ClusterLifecyclePlan sample = ClusterLifecycle.resetNode("local-preview", "local", true, false);
            NodeResult result = NodeResult.pass(ctx.nodeId())
                    .assertion("all_targets_covered", cases == TARGETS.size() * 3)
                    .assertion("normal_reset_runs", normalRuns)
                    .assertion("just_provisioned_reset_skips", justProvisionedSkips)
                    .assertion("already_reset_skips", alreadyResetSkips)
                    .publish("Runtime", "jbang")
                    .publish("ResetCases", Integer.toString(cases));
            for (Map.Entry<String, String> entry : sample.published().entrySet()) {
                result.publish("Reset" + entry.getKey(), entry.getValue());
            }
            return result;
        });
    }
}
